wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/SyncUsersActionHandler.swift

```diff
@@ -0,0 +1,124 @@
+////
+// Wire
+// Copyright (C) 2023 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireDataModel
+
+// rename to describe v4
+class SyncUsersActionHandler: ActionHandler<SyncUsersAction> {
+
+    private let userProfileProcessor = UserProfilePayloadProcessor()
+
+    // MARK: - Request
+
+    struct RequestPayload: Codable, Equatable {
+
+        let qualified_users: [QualifiedID]
+
+    }
+
+    private func markUserProfilesAsUnavailable(_ users: Set<QualifiedID>) {
+        for qualifiedID in users {
+            let user = ZMUser.fetch(with: qualifiedID.uuid, domain: qualifiedID.domain, in: context)
+            user?.isPendingMetadataRefresh = true
+            user?.needsToBeUpdatedFromBackend = false
+        }
+    }
+
+    override func request(
+        for action: SyncUsersAction,
+        apiVersion: APIVersion
+    ) -> ZMTransportRequest? {
+        var action = action
+
+        switch apiVersion {
+        case .v0, .v1, .v2, .v3:
+            action.fail(with: .endpointUnavailable)
+            return nil
+
+        case .v4, .v5:
+            guard
+                let payloadData = RequestPayload(qualified_users: action.qualifiedID).payloadString()
+            else {
+                action.fail(with: .failedToEncodeRequestPayload)
+                return nil
+            }
+
+            return ZMTransportRequest(
+                path: "/list-users",
+                method: .post,
+                payload: payloadData as ZMTransportData,
+                apiVersion: apiVersion.rawValue
+            )
+        }
+    }
+
+    // MARK: - Response
+
+    struct ResponsePayload: Codable {
+
+        let qualified_user_map: Payload.UserProfiles
+
+    }
+
+    override func handleResponse(
+        _ response: ZMTransportResponse,
+        action: SyncUsersAction
+    ) {
+        guard let apiVersion = APIVersion(rawValue: response.apiVersion) else {
+            return
+        }
+
+        var action = action
+
+        switch apiVersion {
+        case .v0, .v1, .v2, .v3:
+            return
```

Comment:
We should still fail the action in this case.

Reply:
Done!